controllers/volumereplicationgroup_controller.go

```diff
@@ -1029,74 +1146,313 @@ func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName, state volr
 	return nil
 }
 
-func (v *VRGInstance) checkVRStatus(volRep *volrep.VolumeReplication) error {
-	// When the generation in the status is updated, VRG would get a reconcile
-	// as it owns VolumeReplication resource.
-	if volRep.Generation != volRep.Status.ObservedGeneration {
-		v.log.Info("Generation from the resource and status not same")
+func (v *VRGInstance) updateVRGStatus(updateConditions bool) error {
+	v.log.Info("Updating VRG status")
 
-		return nil
+	if updateConditions {
+		v.updateVRGConditions()
+	}
+
+	v.updateStatusState()
+
+	v.instance.Status.ObservedGeneration = v.instance.Generation
+	v.instance.Status.LastUpdateTime = metav1.Now()
+
+	if err := v.reconciler.Status().Update(v.ctx, v.instance); err != nil {
+		v.log.Info(fmt.Sprintf("Failed to update VRG status (%s/%s)",
+			v.instance.Name, v.instance.Namespace))
+
+		return fmt.Errorf("failed to update VRG status (%s/%s)", v.instance.Name, v.instance.Namespace)
+	}
+
+	v.log.Info(fmt.Sprintf("Updated VRG Status %+v", v.instance.Status))
+
+	return nil
+}
+
+func (v *VRGInstance) updateStatusState() {
+	vrgCondition := findCondition(v.instance.Status.Conditions, VRGConditionAvailable)
+	if vrgCondition == nil {
+		v.log.Info("Failed to find the Available condition in status")
+
+		return
+	}
+
+	StatusState := getStatusStateFromSpecState(v.instance.Spec.ReplicationState)
+
+	// update Status.State to reflect the state in spec
+	// only after successful transition of the resource
+	// (from primary->secondary or vise versa). That
+	// successful completion of transition can be seen
+	// in vrgCondition.Status being set to True.
+	if vrgCondition.Status == metav1.ConditionTrue {
```

Comment:
On errors we need to clear the status.state. For example we set state to Primary say in generation 1, then we fail to move to Secondary in generation 2, at which point the state would remain Primary, but ObGen would be 2. Now, if the user changed the spec.state to Primary again, then the status would read Primary (from gen 1) and ObGen 3, from latest generation.

I think we need to reset status.state if we start a new generation reconciliation, so that the status.state reads (say, unknown) and obgen is latest when we exit the loop.

Reply:
Just to ensure I understand this correctly, what you are suggesting is to have one more check where if vrgCondition.Status != metav1.ConditionTrue and the reason is Error, then explicitly set the status.state to "unknown" to indicate that some error has happened?

Reply:
Ack! We set status.Conditions always in code calling this routine. The final status update should consider resetting status.state to "something" when there are errors, and possibly unknown is a good option to use.

So overall, yes, if condition is false, update status.state, so that it does not reflect spec.state as that would be incorrect.

Reply:
I need to revisit the code here before marking this resolved. (noting the same)

Reply:
So if VRG condition `VRGConditionAvailable` is in `VRGProgressing` state then `Status.State` is not updated, but `Status.ObservedGeneration` would be the latest. So, `Status.State` maybe stale as per current generation, the real check is to use the `Status.Condition` and not other status fields at present. I am wondering why we need the other status fields in that case?

IOW, obGen, time and state are reflected in the condition and that is sufficient?

Anyway, something to look at, should not stop this patch but tracked separately.

Reply:
Agree with you and even I have the same opinion i.e. do we really need vrg.Status.State and vrg.Status.ObservedGeneration fields. That is why I chose not to rely on status.State in the unittests and just used the vrg conditions to check the success/failure.

I can take it up as a  next action item (along with the suggestion you made for type usage in protectedPVCMap). If you think it would be better to make those changes as part of this PR itself, I can refresh this PR with necessary changes. As of now, having those fields in status should not impact unless the unittests of DRPC (or some other script) checks status.State field.

Reply:
Let's leave it as is for now. @BenamarMk may be integrating and also needs to take a look at relying on conditions (if not already so).

IOW, we can do this as part of a subsequent PR.